### DISCUSSION TITLE
Command decorators: A "complete" guide

### DISCUSSION THREAD

**Discussion by jacueblol** [Patribots4738/Crescendo2024#21] [Announcements]
What is a Command or CommandBase? 

A Command is a class in WPILib that takes in a Runnable and adds it to a running queue or stack of all of the commands in the program. A CommandBase is, in simple terms, an implementation of a Command with more methods and modifiers. You can execute a specific command or the stack as a whole with the .schedule() modifier.

But what is a Command decorator?

A Command decorator is a method from the Command or, more likely, a Command Base class that has just been returned. 

A Command modifier will modify the Command class that it is attached to. It is possible to layer Commands by putting one Command inside of another; however, if a modifier is applied to the outside Command, it is not applied to the root or the inside Command. Let’s look at an example of this in #4 

Below is the previous version of the code, with a layered Command…

    public void onDisabled() {
        Commands.run(this::update).ignoringDisable(true).schedule();
    }
    ...
    public Command update() {
        return Commands.runOnce(() -> FieldConstants.ALLIANCE = DriverStation.getAlliance());
    }

The code above does not work. The reason for this error is because the Command modifiers being applied to the outside command, not the original or root command: “update()”.

![image](https://github.com/Patribots4738/Swerve-Command-Based/assets/77470805/f3eac8ec-bd5c-4a91-9773-abb038e5b106)

This error forces the “onDisabled()” method to run the Commands.run() method. However, this logic does not apply to the inside command, “update(),” which is what we want to happen. 

    public void onDisabled() {
        update().ignoringDisable(true).repeatedly().schedule();
    }
    ...
    public Command update() {
        return Commands.runOnce(() -> {
            FieldConstants.ALLIANCE = DriverStation.getAlliance();
        });
    }

**Comment by GalexY727** [comment]
I would like to have a discussion as to if things like this should be done iteratively or not, because we are given a disabledperiodic after all

**Reply by jacueblol** [reply]
I believe that we should stick with command based, however it would be interesting to hear everyone's thoughts on the topic. 

I believe an option is going half in and half out. Doing part command based, and part iterative.

I'll bring it up at the scrum today.

**Reply by GalexY727** [reply]
We did not in fact bring it up in scrum 5 days ago
